common/src/character_controller.rs

```diff
@@ -91,6 +81,46 @@ impl CharacterControllerPass<'_> {
         }
     }
 
+    /// Updates the position based on the given average velocity while handling collisions. Also updates the velocity
+    /// based on collisions that occur.
+    fn apply_velocity(&mut self, estimated_average_velocity: &na::Vector3<f32>) {
+        // To prevent an unbounded runtime, we only allow a limited number of collisions to be processed in
+        // a single step. If the player encounters excessively complex geometry, it is possible to hit this limit,
+        // in which case further movement processing is delayed until the next time step.
+        const MAX_COLLISION_ITERATIONS: u32 = 5;
+
+        let mut expected_displacement = estimated_average_velocity * self.dt_seconds;
+        let mut active_normals = Vec::<na::UnitVector3<f32>>::with_capacity(3);
```

Comment:
Is accumulating and simultaneously solving multiple normals necessary? Would it be sufficient to instead treat each collision independently, each time subtracting the portion of the remaining `expected_displacement` that would move into that individual normal?

Reply:
This causes problems if the player moves across an inner edge with an acute angle. If each normal is treated one-at-a-time, the first surface will push the player towards the second surface, and the second surface will then push the player back towards the first surface, an infinite loop that makes forward progress impossible.

Reply:
Ah, missed the `retain`.

Reply:
Not an infinite loop, since the maximum iterations are bounded, and motion perpendicular to the axis of the angle should still be possible.

Reply:
Right, the loop wouldn't be infinite, but if we reach the bound, the behavior usually isn't ideal.

The issue is that the motion vector after being affected by one normal at a time (without any consideration of the other normal) will always point towards at least one of the walls of the wedge.

Without this extra logic, it would go as follows in Euclidean space: Some forward progress is made originally before hitting the first wall. Then, some more forward progress is made, sliding against the first wall, before hitting the second wall. However, now that the player is now touching both walls at once, as long as the `expected_displacement` faces one of the two walls, a collision is always reported immediately without any forward movement (`tanh_distance == 0`), so the player gets stuck.

In hyperbolic space, since the wall diverges a bit from the player, the player wouldn't quite hit both walls at once, so some forward progress would be made in every step, but that forward progress would exponentially decay, preventing collision from getting resolved before `MAX_COLLISION_ITERATIONS`. At the scales we normally deal with, it basically feels exactly like the Euclidean case.

Reply:
> as long as the `expected_displacement` faces one of the two walls, a collision is always reported immediately without any forward movement

Why would `expected_displacement` face either wall, after the component facing them was subtracted out in the first two iterations? Isn't it at that point left necessarily either perpendicular to or away from them?

Reply:
Here's an image which demonstrates the issue. It shows 4 collision steps of the sphere approaching a wedge between two planes. There is movement in the z-direction that we don't see in this diagram because I made it 2D for simplicity. The z-component of these vectors never get redirected.

![image](https://user-images.githubusercontent.com/5403380/222937913-737f6224-28ad-4ff0-9e70-340e59890d69.png)

In each of the images, the sphere initially has an expected displacement equal to the blue vector, and subtracting out the wall normal turns it into the green vector. The green vector in the previous image corresponds to the blue vector in the next image.

In stage 1, the sphere just hit one of the two walls, causing it to be redirected to slide against this wall.

In stage 2, the sphere just hit the other wall, causing it to get redirected downwards again. I draw the vectors starting on the relevant contact point to help better illustrate which wall is triggering this redirection, but I could have just as easily drawn them anywhere else.

In stage 3, the sphere immediately hits the lower wall without moving, causing the expected displacement to redirect upwards again. It cannot make any progress in the z-direction because it immediately hit a wall.

In stage 4, the blue vector is facing the same vector as in stage 2, so we get a repeat. Stages 2 and 3 effectively repeat forever until we reach the maximum iteration count, causing the sphere to get stuck.

The main issue is that the walls are not perpendicular to each other, so if we subtract out each wall's normal one at a time, they'll always undo each other's work. Having both normals act at once lets us handle this case correctly.

Reply:
Good illustration, thanks! I wonder why I've never ran into this issue in other projects... certainly we're more likely here than most to be walking into/over/along weird angles.